crates/autopilot/src/decoded_settlement.rs

```diff
@@ -226,69 +236,72 @@ impl DecodedSettlement {
                     flags: trade.8.into(),
                     executed_amount: trade.9,
                     signature: trade.10,
-                })
-                .collect(),
+                };
+                match trade.compute_uid(domain_separator, &tokens) {
+                    Ok(order_uid) => {
+                        trade.order_uid = order_uid;
+                        Some(trade)
+                    }
+                    Err(err) => {
+                        tracing::error!(
+                            ?err,
+                            ?trade,
+                            "failed to calculate order uid, we don't know which order this trade \
+                             belongs to"
+                        );
+                        None
+                    }
+                }
+            })
+            .collect();
+        Ok(Self {
+            tokens,
+            clearing_prices,
+            trades,
             interactions: interactions.map(|inner| inner.into_iter().map(Into::into).collect()),
             metadata: metadata.try_into().ok().map(Bytes),
         })
     }
 
     /// Returns the total surplus denominated in the native asset for the
     /// solution.
-    pub fn total_surplus(&self, external_prices: &ExternalPrices) -> U256 {
-        self.trades.iter().fold(0.into(), |acc, trade| {
-            acc + match surplus(trade, &self.tokens, &self.clearing_prices, external_prices) {
-                Some(surplus) => surplus,
-                None => {
-                    tracing::warn!("possible incomplete surplus calculation");
-                    0.into()
-                }
-            }
-        })
-    }
-
-    /// Returns fees for all trades.
-    pub fn all_fees(
+    pub fn total_surplus(
         &self,
         external_prices: &ExternalPrices,
-        domain_separator: &DomainSeparator,
-    ) -> Vec<Fees> {
+        surplus_exempt_order_uids: &HashSet<OrderUid>,
```

Comment:
I'm not sure this is the best way to handle this. You basically force the caller of total_surplus to think about and remember to fetch the original auction and excluded jit orders. What purpose do jit orders serve on this struct in the first place? Would it make sense to pass in the original auction into the constructor and do the separation there instead of carrying this around as an extra argument?

Reply:
> What purpose do jit orders serve on this struct in the first place?

Not sure I understand the question. When computing the score we need to be aware about JIT orders in some way since CoW AMM JIT orders are supposed to receiver surplus.
The question is in what what way should the struct learn about which orders are allowed to receive surplus.

Reply:
I don't think the auction can be a part of the `DecodedSettlement` struct since we first need to decode a settlement and extract an `auction_id` to fetch auction JSON from the `solver_competitions` table. Another option is introducing a new dto struct containing a decoded settlement with an auction.

Reply:
Let's discuss this in a more detail here: https://github.com/cowprotocol/services/issues/2723

For now, I am fine implementing it here in a "just make it work" manner :), since this file will be dropped anyway quite soon.

Reply:
> I don't think the auction can be a part of the DecodedSettlement struct since we first need to decode a settlement and extract an auction_id to fetch auction JSON from the solver_competitions table.

I think we could call a settlement as not decodable if we aren't able to create it **and** fetch the related auction.